empyrical/perf_attrib.py

```diff
@@ -0,0 +1,409 @@
+        - Example:
+                        momentum  reversal
+            2017-01-01  0.002779 -0.005453
+            2017-01-02  0.001096  0.010290
+
+    factor_loadings : pd.DataFrame
+        Factor loadings for all days in the date range, with date and ticker as
+        index, and factors as columns.
+        - Example:
+                               momentum  reversal
+            dt         ticker
+            2017-01-01 AAPL   -1.592914  0.852830
+                       TLT     0.184864  0.895534
+                       XOM     0.993160  1.149353
+            2017-01-02 AAPL   -0.140009 -0.524952
+                       TLT    -1.066978  0.185435
+                       XOM    -1.798401  0.761549
+
+    transactions : pd.DataFrame, optional
+        Executed trade volumes and fill prices. Used to check the turnover of
+        the algorithm. Default is None, in which case the turnover check is
+        skipped.
+
+        - One row per trade.
+        - Trades on different names that occur at the
+          same time will have identical indicies.
+        - Example:
+            index                  amount   price    symbol
+            2004-01-09 12:18:01    483      324.12   'AAPL'
+            2004-01-09 12:18:01    122      83.10    'MSFT'
+            2004-01-13 14:12:23    -75      340.43   'AAPL'
+
+    pos_in_dollars : bool
+        Flag indicating whether `positions` are in dollars or percentages
+        If True, positions are in dollars.
+
+    Returns
+    -------
+    tuple of (risk_exposures_portfolio, perf_attribution)
+
+    risk_exposures_portfolio : pd.DataFrame
+        df indexed by datetime, with factors as columns
+        - Example:
+                        momentum  reversal
+            dt
+            2017-01-01 -0.238655  0.077123
+            2017-01-02  0.821872  1.520515
+
+    perf_attribution : pd.DataFrame
+        df with factors, common returns, and specific returns as columns,
+        and datetimes as index
+        - Example:
+                        momentum  reversal  common_returns  specific_returns
+            dt
+            2017-01-01  0.249087  0.935925        1.185012          1.185012
+            2017-01-02 -0.003194 -0.400786       -0.403980         -0.403980
+    """
+    (returns,
+     positions,
+     factor_returns,
+     factor_loadings) = _align_and_warn(returns,
+                                        positions,
+                                        factor_returns,
+                                        factor_loadings,
+                                        transactions,
+                                        pos_in_dollars=pos_in_dollars)
+
+    # Note that we convert positions to percentages *after* the checks
+    # above, since get_turnover() expects positions in dollars.
+    positions = _stack_positions(positions, pos_in_dollars=pos_in_dollars)
+
+    risk_exposures_portfolio = compute_exposures(positions,
```

Comment:
I think the only logic we actually need for perf attribution in zipline is the code from here through the end of this function.  In general, my mental model is that, as much as possible, empyrical should be just the "bare bones" calculations.  What do you think about carving out just the logic of "Given returns, positions, factor_returns, and factor_loadings, compute perf attribution" and leave the logic of "display warnings about missing/unreasonable data" to pyfolio?

Reply:
I can do that. Are we assuming that anything calling into empyrical has its inputs already correctly formatted?

Reply:
only concern is that it'll result in duplicating some of the logic to make indexes match up, etc., before perf attrib gets called

Reply:
I think so.  Otherwise we're emposing a performance penalty on a caller that already has well-aligned data.  In PyFolio, the expected user is running a tearsheet interactively in a notebook, so the extra overhead of doing data validation isn't a big deal, but I don't think we want to make the assumption that those checks are always an acceptable overhead.

Reply:
Also, in zipline, there are cases where we won't have all the transactions in memory to use for turnover (e.g., in live algorithms, we don't want to reload all the historical transactions).